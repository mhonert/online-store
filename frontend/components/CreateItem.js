import React, {useState} from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import Router from 'next/router';

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
    $title: String!,
    $description: String!,
    $price: Int!,
    $image: String,
    $largeImage: String
    ) {
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`

const CreateItem = () => {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');
  const [image, setImage] = useState('');
  const [largeImage, setLargeImage] = useState('');
  const [price, setPrice] = useState(0);

  const handleSubmit = async (e, createItem) => {
    // Stop the for from submitting
    e.preventDefault();

    // call the mutation
    const result = await createItem();

    // change them to the single item page
    Router.push({
      pathname: '/item',
      query: {id: result.data.createItem.id}
    });
  }

  return (
    <Mutation mutation={CREATE_ITEM_MUTATION} variables={{title, description, image, largeImage, price}}>
      {(createItem, { loading, error }) => (
        <Form onSubmit={e => handleSubmit(e, createItem)}>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
            <label>Title
              <input type="text" name="title"
                     placeholder="Title" required
                     value={title} onChange={e => setTitle(e.target.value)} />
            </label>
            <label>Price
              <input type="number" name="price"
                     placeholder="Price" required
                     value={price} onChange={e => setPrice(e.target.value)} />
            </label>
            <label>Description
              <textarea type="text" name="description"
                        placeholder="Enter a Description" required
                        value={description} onChange={e => setDescription(e.target.value)} />
            </label>
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
    </Mutation>
  );
};

export default CreateItem;
