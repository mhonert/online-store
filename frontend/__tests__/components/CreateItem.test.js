import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';

import { MockedProvider } from 'react-apollo/test-utils';
import React from 'react';
import Router from 'next/router';
import CreateItem, { CREATE_ITEM_MUTATION } from '../../components/CreateItem';
import { fakeItem } from '../../lib/testUtils';

// mock runtime configuration which is usually loaded by next.js from the config.js file,
// but is unavailable during testing
jest.mock('next/config', () => () => ({
  publicRuntimeConfig: { stripePublishableKey: 'test' }
}));

const testImage = 'https://localhost/test.jpg';

global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: testImage,
    eager: [{ secure_url: testImage }]
  })
});

Router.router = {
  push() {},
  prefetch() {}
};

describe('<CreateItem/>', () => {
  it('Renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });

  it('Uploads a file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    const input = wrapper.find('input[type="file"]');
    input.simulate('change', { target: { files: ['test.jpg'] } });
    await wait();
    wrapper.update();

    expect(wrapper.find('img').prop('src')).toBe(testImage);
    expect(global.fetch).toBeCalled();
  });

  it('Creates and item when the form is submitted', async () => {
    const item = fakeItem();
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            description: item.description,
            image: '',
            largeImage: '',
            price: item.price
          }
        },
        result: {
          data: {
            createItem: {
              ...item,
              __typename: 'Item'
            }
          }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem/>
      </MockedProvider>
    );

    // simulate form input
    simulateInput(wrapper, 'title', item.title);
    simulateInput(wrapper, 'price', item.price, 'number');
    simulateInput(wrapper, 'description', item.description);

    // mock the router
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');

    await wait(50);
    wrapper.update();

    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: "/item",
      query: {
        id: item.id
      }
    });
  });
});

const simulateInput = (wrapper, name, value, type) => {
  wrapper
    .find(`#${name}`)
    .simulate('change', { target: { name, value, type } });
};
