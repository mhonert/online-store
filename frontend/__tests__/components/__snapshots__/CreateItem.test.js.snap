// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> Renders and matches snapshot 1`] = `
<form
  className="Form-sc-3ek8g3-0 kxkTjx"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label>
      Image
      <input
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label>
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label>
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label>
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a Description"
        required={true}
        type="text"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
